### DISCUSSION TITLE
Unable to use pipes with FITS images

### DISCUSSION THREAD

**Discussion by ewelot** [libvips/libvips#2096] [Q&A]
Similar to [this (already solved) issue](https://github.com/libvips/libvips/issues/1916) there seems to be no support for reading/writing FITS images to/from stdin/stdout. Is this a general limitation for FITS files or just a currently missing implementation detail?

Thanks for help,
Thomas

**Comment by jcupitt** [comment]
Hi @ewelot, cfitsio only works with files on disc, as far as I know:

https://heasarc.gsfc.nasa.gov/docs/software/fitsio/c/c_user/node35.html

You'd need to read the whole pipe to a temp file, then open that.

**Comment by ewelot** [comment]
I am pretty sure cfitsio is able to read from stdin (and support e.g. http streams). Citing from the page you referenced:
```
...
Programs can read the input file from the 'stdin' file stream if a dash character ('-') is given as the filename. Files can also be opened over the network using FTP or HTTP protocols by supplying the appropriate URL as the filename. The HTTPS and FTPS protocols are also supported if the CFITSIO build includes the libcurl library. (If the CFITSIO 'configure' script finds a usable libcurl library on your system, it will automatically be included in the build.) 
...
The output file will be written to the 'stdout' file stream if a dash character ('-') or the string 'stdout' is given as the filename. Similarly, '-.gz' or 'stdout.gz' will cause the file to be gzip compressed before it is written out to the stdout stream. 
...
```

**Reply by jcupitt** [reply]
Sorry @ewelot, I never replied to this :(

cfitsio (as far as I know) does not expose this functionality via an API, it only has this magic filename trick.

It would need to have some API that let the user pass in a read callback (like libjpeg, tiff, png etc.), or perhaps use a state machine (like libjxl, libwebp, etc.).

Maybe it is possible? Someone brave would need to sit down and look though the cfitsio sources.

**Comment by ewelot** [comment]
Thanks for the reply!
Unfortunately I dont have the skills to help solving this issue.
I will try to work around by avoiding pipes as much as possible.

Anyway, thanks for your support
Thomas

> Am Wed, 10 Nov 2021 04:45:09 -0800
> schrieb John Cupitt ***@***.***>:
>
> Sorry @ewelot, I never replied to this :(
> 
> cfitsio (as far as I know) does not expose this functionality via an API, it only has this magic filename trick.
> 
> It would need to have some API that let the user pass in a read callback (like libjpeg, tiff, png etc.), or perhaps use a state machine (like libjxl, libwebp, etc.).
> 
> Maybe it is possible? Someone brave would need to sit down and look though the cfitsio sources.
>
